google/cloud/universe_domain.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/universe_domain.h"
+#include "google/cloud/credentials.h"
+#include "google/cloud/internal/populate_common_options.h"
+#include "google/cloud/internal/unified_rest_credentials.h"
+#include "google/cloud/universe_domain_options.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+template <typename T, typename U = typename std::remove_reference_t<
+                          decltype(*typename T::Type())>>
+std::unique_ptr<U> GetPolicyClone(Options const& options) {
+  std::unique_ptr<U> policy;
+  if (options.has<T>()) {
+    auto const& policy_option = options.get<T>();
+    if (policy_option) return policy_option->clone();
+  }
+  return policy;
+}
+
+}  // namespace
+
+StatusOr<Options> AddUniverseDomainOption(ExperimentalTag, Options options) {
+  if (!options.has<UnifiedCredentialsOption>()) {
+    options.set<UnifiedCredentialsOption>(
+        MakeGoogleDefaultCredentials(internal::MakeAuthOptions(options)));
+  }
+
+  auto universe_domain = GetUniverseDomain(
+      ExperimentalTag{}, *options.get<UnifiedCredentialsOption>(),
+      GetPolicyClone<internal::UniverseDomainRetryPolicyOption>(options),
+      GetPolicyClone<internal::UniverseDomainBackoffPolicyOption>(options));
+  if (!universe_domain) return std::move(universe_domain).status();
+  options.set<internal::UniverseDomainOption>(*std::move(universe_domain));
+  return options;
+}
+
+StatusOr<std::string> GetUniverseDomain(
+    ExperimentalTag, Credentials const& credentials,
+    std::unique_ptr<internal::UniverseDomainRetryPolicy> retry_policy,
+    std::unique_ptr<BackoffPolicy> backoff_policy) {
+  Options retry_options;
+  if (retry_policy) {
+    retry_options.set<internal::UniverseDomainRetryPolicyOption>(
+        std::move(retry_policy));
+  }
+  if (backoff_policy) {
+    retry_options.set<internal::UniverseDomainBackoffPolicyOption>(
+        std::move(backoff_policy));
+  }
```

Comment:
seems like this can go away if we just ask for the `Options`. as well as the `GetPolicyClone` above.